feat: add title field and metadata handling for ZIM file creation

nicolapace opened the pull request:
Fixes #677 

I added a `Title` field in the Zim creation page. This field is automatically filled with the `Selection Title` reduced to 30 characters. In this way users can change it if they want. This field is required and is of max 30 chars. (I also added a char counter for both the title and the description).

I also added backend validation for the size of zim metadata (title (30) and description (80) as per https://wiki.openzim.org/wiki/Metadata).

I added new unit test for the new "title field" and the validation, for both the backend and the frontend. I also fixed some unit tests and functions to support the new `zimtitle` field.

Here is the Zim creation page with the new field:

![Screenshot](https://github.com/user-attachments/assets/309046a2-43d5-4b6f-ae63-a6be476cec2c)

audiodude commented:
> @nicolapace Thank you for your PR! But the unity for counting is not char, but lexem. There is a javascript/typescript implementation of such a counter in openzim/mwoffliner.

https://github.com/openzim/wp1/issues/677#issuecomment-2807939549

What is a lexem and how does it compare to a grapheme? I feel like I'm going to have to read a Chomsky book to implement this PR. :upside_down_face:

audiodude reviewed (changes requested):
This looks really good, great work!

I think the handling of the error conditions on the backend needs to be improved however.

kelson42 commented:
@audiodude I meant grapheme, sorry.

nicolapace commented on wp1/zimfarm.py:
```diff
@@ -121,7 +125,23 @@ def get_webhook_url():
+def validate_zim_metadata(title=None, description=None):
+  """Validate ZIM metadata fields against length limits."""
+  errors = []
+  print(f"Validating ZIM metadata: title='{title}', description='{description}'")
```
Sure :)

nicolapace commented:
@audiodude @kelson42 @benoit74 Thanks for the feedback and the suggestions!

I pushed a new commit that addresses the discussed issues, in particular:
 - I am now using `split-by-grapheme` npm package and `regex` python module to check for num. of graphemes in both frontend and backend as suggested by @kelson42 and @benoit74
 - as suggested by @audiodude I changed the `validate_zim_metadata()` function to Raise exceptions directly and I created new exceptions
 - I also added validation for the `long_description` field as described in https://wiki.openzim.org/wiki/Metadata#Graphemes
 - I disable the submit button if inputs are not valid
 - I changed unit tests to support the new changes and added some new unit tests (for the long description validation and the handling of graphemes)

(I pushed a new commit instead of amending the last one for better clarity. We can squash them later if you prefer.)

nicolapace commented on wp1/zimfarm.py:
```diff
@@ -121,7 +125,23 @@ def get_webhook_url():
+    errors.append(f"Description exceeds maximum length: {ZIM_DESCRIPTION_MAX_LENGTH} characters.")
+    
+  if errors:
+    return False, " ".join(errors)
```
Thanks! And yes, totally agree, makes sense. We handle user feedback on the frontend and just return errors on the backend calls. I’ve switched to raising specific exceptions like `InvalidZimTitleError` and `InvalidZimDescriptionError` (as subclasses of `ZimFarmError`) instead of using status tuples. (I've also added `InvalidZimLongDescriptionError` for `long_description` related errors). Thanks again for the feedback!

audiodude commented on wp1-frontend/src/components/ZimFile.vue:
```diff
@@ -72,13 +72,15 @@
               />
-              <small class="form-text text-muted">{{ zimTitle.length }}/30 characters</small>
+              <small class="form-text" :class="{'text-muted': graphemeCount(zimTitle) < maxTitleLength, 'text-warning': graphemeCount(zimTitle) === maxTitleLength}">
+                {{ graphemeCount(zimTitle) }}/{{ maxTitleLength }} graphemes{{ graphemeCount(zimTitle) === maxTitleLength ? ' (max reached)' : '' }}
```
This is a lot of information/processing to be inside of a {{ ... }} block. Please refactor to a helper function. Specifically the last part with the ternary operator. I would  a full on `displayGraphemeLimitText()` helper.

audiodude commented on wp1-frontend/src/components/ZimFile.vue:
```diff
@@ -87,13 +89,15 @@
               />
-              <small class="form-text text-muted">{{ description.length }}/80 characters</small>
+              <small class="form-text" :class="{'text-muted': graphemeCount(description) < maxDescriptionLength, 'text-warning': graphemeCount(description) === maxDescriptionLength}">
+                {{ graphemeCount(description) }}/{{ maxDescriptionLength }} graphemes{{ graphemeCount(description) === maxDescriptionLength ? ' (max reached)' : '' }}
```
Same feedback here.

audiodude commented on wp1-frontend/src/components/ZimFile.vue:
```diff
@@ -314,6 +341,16 @@ export default {
         this.status === 'ENDED'
       );
     },
+    isLongDescriptionInvalid: function() {
```
nit: prefer booleans that test for positive attributes. So `isLongDescriptionValid()` and then use `!isLongDescriptionValid()` in your test

audiodude commented on wp1/zimfarm_test.py:
```diff
@@ -408,11 +408,27 @@ def test_schedule_zim_file_task_create_raises(self, get_params_mock,
+                             mock_requests):
+    redis = MagicMock()
+    s3 = MagicMock()
+    get_params_mock.return_value = {'name': 'bar'}
```
Optional: I don't think you need to set up all the response values for the mocks here. We just need to send in any mock as a requirement for the function, so that it gets to the point where it can raise the exception.

I recognize that other `test_invalid_foo` methods in this test do that, so it would be optional but nice to clean it up, especially as we are adding a few more methods of this type. Alternately you could simply fix it in the new methods and ignore the existing calls.

audiodude commented on wp1/zimfarm_test.py:
```diff
@@ -423,7 +439,24 @@ def test_schedule_zim_file_long_name(self, get_params_mock, get_token_mock,
+                             mock_requests):
+    redis = MagicMock()
+    s3 = MagicMock()
+    get_params_mock.return_value = {'name': 'bar'}
```
Same comment here.

audiodude commented on wp1/zimfarm_test.py:
```diff
@@ -439,12 +472,60 @@ def test_schedule_zim_file_too_long_description(self, get_params_mock, get_token
+                             mock_requests):
+    redis = MagicMock()
+    s3 = MagicMock()
+    get_params_mock.return_value = {'name': 'bar'}
```
Same

audiodude reviewed (changes requested):
Looking good, just a few small comments.

audiodude reviewed (approved):
Great work!
